modules/dnn/src/net_impl.cpp

```diff
@@ -1426,6 +1426,7 @@ void Net::Impl::setInput(InputArray blob, const String& name, double scalefactor
 
     LayerData& ld = layers[pin.lid];
     const int numInputs = std::max(pin.oid + 1, (int)ld.requiredOutputs.size());
+    ld.dtype = blob_.type();
```

Comment:
This modification allows `net` to modify the type of input layer according to the input blob type. Without it, the type of input layer is always `CV_32F`. 
Not sure this change is safe. @asmorkalov, @vpisarev.